# --- T2-COPYRIGHT-NOTE-BEGIN ---
# This copyright note is auto-generated by ./scripts/Create-CopyPatch.
# 
# T2 SDE: package/.../uclibc/scanf-aflag.patch
# Copyright (C) 2007 The T2 SDE Project
# 
# More information can be found in the files COPYING and README.
# 
# This patch file is dual-licensed. It is available under the license the
# patched project is licensed under, as long as it is an OpenSource license
# as defined at http://www.opensource.org/ (e.g. BSD, X11) or under the terms
# of the GNU General Public License as published by the Free Software
# Foundation; either version 2 of the License, or (at your option) any later
# version.
# --- T2-COPYRIGHT-NOTE-END ---

--- uClibc-20060830/libc/stdio/_scanf.c	2006-08-30 09:10:17.000000000 +0200
+++ uClibc-0.9.28/libc/stdio/_scanf.c	2006-11-01 21:19:32.000000000 +0100
@@ -105,12 +105,11 @@
 #endif
 
 #ifdef __UCLIBC_HAS_SCANF_GLIBC_A_FLAG__
-#ifdef L_vfscanf
-/* only emit this once */
-#warning Forcing undef of __UCLIBC_HAS_SCANF_GLIBC_A_FLAG__ until implemented!
-#endif
+#ifdef L_vfwscanf
+#warning Forcing undef of __UCLIBC_HAS_SCANF_GLIBC_A_FLAG__ until implemented for wchar!
 #undef __UCLIBC_HAS_SCANF_GLIBC_A_FLAG__
 #endif
+#endif
 
 #undef __STDIO_HAS_VSSCANF
 #if defined(__STDIO_BUFFERS) || !defined(__UCLIBC_HAS_WCHAR__) || defined(__UCLIBC_HAS_GLIBC_CUSTOM_STREAMS__)
@@ -934,7 +933,6 @@
 			int p_m_spec_chars = p - spec_chars;
 
 #ifdef __UCLIBC_HAS_SCANF_GLIBC_A_FLAG__
-#error implement gnu a flag
 			if ((*p == 'a')
 				&& ((psfs->fmt[1] == '[') || ((psfs->fmt[1]|0x20) == 's'))
 				) {		/* Assumes ascii for 's' and 'S' test. */
@@ -1187,6 +1185,12 @@
 
 	int i;
 
+#ifdef __UCLIBC_HAS_SCANF_GLIBC_A_FLAG__
+#define MALLOC_CHUNK_SIZE	64
+	unsigned int  alen, aoffset;
+	unsigned char *abuf;
+#endif
+
 #ifdef __UCLIBC_MJN3_ONLY__
 #warning TODO: Fix MAX_DIGITS.  We do not do binary, so...!
 #endif
@@ -1411,6 +1415,15 @@
 				b = (psfs.store ? ((unsigned char *) psfs.cur_ptr) : buf);
 				fail = 1;
 			
+#ifdef __UCLIBC_HAS_SCANF_GLIBC_A_FLAG__
+				if (psfs.flags & FLAG_MALLOC) {
+					alen = MALLOC_CHUNK_SIZE;
+					if ((abuf = (unsigned char *) malloc(alen)) == NULL) {
+						goto NEXT_FMT;
+					}
+					b = abuf;
+				}
+#endif
 
 				if (psfs.conv_num == CONV_c) {
 					if (sc.width == INT_MAX) {
@@ -1418,6 +1431,16 @@
 					}
 
 					while (__scan_getc(&sc) >= 0) {
+#ifdef __UCLIBC_HAS_SCANF_GLIBC_A_FLAG__
+						aoffset = (unsigned int) b - (unsigned int) abuf;
+						if ((psfs.flags & FLAG_MALLOC) && ( aoffset + psfs.store >= alen)) {
+							alen += MALLOC_CHUNK_SIZE;
+							if ((abuf = (unsigned char *) realloc(abuf, alen)) == NULL) {
+								break;
+							}
+							b = abuf + aoffset;
+						}
+#endif
 						zero_conversions = 0;
 						*b = sc.cc;
 						b += psfs.store;
@@ -1433,6 +1456,17 @@
 				if (psfs.conv_num == CONV_s) {
 					/* Yes, believe it or not, a %s conversion can store nuls. */
 					while ((__scan_getc(&sc) >= 0) && !isspace(sc.cc)) {
+#ifdef __UCLIBC_HAS_SCANF_GLIBC_A_FLAG__
+						aoffset = (unsigned int) b - (unsigned int) abuf;
+						if ((psfs.flags & FLAG_MALLOC) && ( aoffset + psfs.store >= alen)) {
+							alen += MALLOC_CHUNK_SIZE;
+							if ((abuf = (unsigned char *) realloc(abuf, alen)) == NULL) {
+								fail = 1;
+								break;
+							}
+							b = abuf + aoffset;
+						}
+#endif
 						zero_conversions = 0;
 						*b = sc.cc;
 						b += psfs.store;
@@ -1493,6 +1527,17 @@
 						if (!scanset[sc.cc]) {
 							break;
 						}
+#ifdef __UCLIBC_HAS_SCANF_GLIBC_A_FLAG__
+						aoffset = (unsigned int) b - (unsigned int) abuf;
+						if ((psfs.flags & FLAG_MALLOC) && ( aoffset + psfs.store >= alen)) {
+							alen += MALLOC_CHUNK_SIZE;
+							if ((abuf = (unsigned char *) realloc(abuf, alen)) == NULL) {
+								fail = 1;
+								break;
+							}
+							b = abuf + aoffset;
+						}
+#endif
 						*b = sc.cc;
 						b += psfs.store;
 						fail = 0;
@@ -1500,6 +1545,12 @@
 				}
 				/* Common tail for processing of %s and %[. */
 
+#ifdef __UCLIBC_HAS_SCANF_GLIBC_A_FLAG__
+				if (psfs.flags & FLAG_MALLOC) {
+					/* store the ptr to allocated string space, or even null */
+					* (unsigned char **) psfs.cur_ptr = abuf;
+				}
+#endif
 				__scan_ungetc(&sc);
 				if (fail) {	/* nothing stored! */
 					goto DONE;
